posthog/hogql/printer.py

```diff
@@ -348,15 +352,40 @@ def visit_compare_operation(self, node: ast.CompareOperation):
         left = self.visit(node.left)
         right = self.visit(node.right)
         if node.op == ast.CompareOperationOp.Eq:
-            if isinstance(node.right, ast.Constant) and node.right.value is None:
-                return f"isNull({left})"
-            else:
+            if isinstance(node.left, ast.Constant) and isinstance(node.right, ast.Constant):
+                return "1" if node.left.value == node.right.value else "0"
+            elif (
+                self.in_join_on or self.dialect == "hogql"
+            ):  # TODO: get this info from the stack (needs clause support)
                 return f"equals({left}, {right})"
-        elif node.op == ast.CompareOperationOp.NotEq:
-            if isinstance(node.right, ast.Constant) and node.right.value is None:
-                return f"isNotNull({left})"
+            elif isinstance(node.right, ast.Constant):
+                if node.right.value is None:
+                    return f"isNull({left})"
+                return f"ifNull(equals({left}, {right}), 0)"
+            elif isinstance(node.left, ast.Constant) and node.left.value is None:
+                if node.left.value is None:
+                    return f"isNull({right})"
+                return f"ifNull(equals({left}, {right}), 0)"
```

Comment:
Is zero safe to use? This will make it so that if you do `property.foo = 0` it will match all events that don't have `foo`. What if we used a somewhat random, unlikely to be found in the wild "sentinel" value like `_h_NULL_`?

Reply:
The 0-s and 1-s here represent booleans. Using `true` or `false` gives an error, so I think there's nothing else we can use:

<img width="1992" alt="image" src="https://github.com/PostHog/posthog/assets/53387/2639dfa1-3413-454e-aa6c-ec5ded2a965f">

And comparing `0 = (1 = 2)` aka `equals(0, equals(1,2)` returns `1`:

<img width="748" alt="image" src="https://github.com/PostHog/posthog/assets/53387/6e818ca5-a4ca-4aaf-9113-75f95d06c221">

Reply:
Ah, right, totally missed that the zero is outside `equals()`